### PR TITLE
Add navigation-fragment module

### PR DESCRIPTION
**Pull request by tfcporciuncula**
Fragment related work for https://github.com/marcellogalhardo/retained/issues/14

The goal here is to expose a new `retainInNavGraph` extension for fragments that can be used to retain instances scoped to a nav graph. In the future we can also to do the same for views (once https://github.com/marcellogalhardo/retained/pull/6 is merged) and compose.

Relevant resources:
- [`Fragment.navGraphViewModels`](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:navigation/navigation-fragment-ktx/src/main/java/androidx/navigation/NavGraphViewModelLazy.kt)
- [`Fragment.hiltNavGraphViewModels`](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:hilt/hilt-navigation-fragment/src/main/java/androidx/hilt/navigation/fragment/HiltNavGraphViewModelLazy.kt)
- [`hiltNavGraphViewModel`(for compose)](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-main:hilt/hilt-navigation-compose/src/main/java/androidx/hilt/navigation/compose/NavHostController.kt)

We need to incorporate this in the sample (or create a new sample) and write tests, but I was thinking we could have those in follow up PRs.

### CODE REVIEW

File: navigation-fragment/src/main/java/dev/marcellogalhardo/retained/navigation/fragment/NavigationFragmentRetainedObject.kt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+ * @see createRetainedObject
+ */
+@OptIn(InternalRetainedApi::class)
+inline fun <reified T: Any> Fragment.retainInNavGraph(
+    @IdRes navGraphId: Int,
+    key: String = T::class.java.name,
+    noinline getDefaultArgs: (() -> Bundle)? = null,
+    noinline createRetainedObject: (RetainedEntry) -> T
+): Lazy<T> {
+    val backStackEntry by lazy(LazyThreadSafetyMode.NONE) { findNavController().getBackStackEntry(navGraphId) }
```
Comment (tfcporciuncula) [inline navigation-fragment/src/main/java/dev/marcellogalhardo/retained/navigation/fragment/NavigationFragmentRetainedObject.kt:35]: Google's implementation use a regular `lazy`, but I honestly can't see a reason for the synchronization here.

File: navigation-fragment/build.gradle [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+afterEvaluate {
+    publishing {
+        publications {
+            release(MavenPublication) {
+                from components.release
+
+                groupId = 'dev.marcellogalhardo'
+                artifactId = 'retained-fragment'
```
Comment (marcellogalhardo) [inline navigation-fragment/build.gradle:50]: Artefact name should be `retained-navigation-fragment` to match the module structure.

**Review by marcellogalhardo — COMMENTED** [review]
Looks good to me! After we fix the artefact ID we can move forward with those changes.

Thank you @tfcporciuncula !

File: navigation-fragment/build.gradle (hunk 2 above)
Comment (tfcporciuncula) [reply navigation-fragment/build.gradle:50]: Oops! Copy paste mistake 🤦‍♂️

**Review by marcellogalhardo — APPROVED** [review]

**Pull request merged**